packages/documentsearch/src/providers/notebooksearchprovider.ts

```diff
@@ -109,13 +117,46 @@ export class NotebookSearchProvider implements ISearchProvider<NotebookPanel> {
       });
     }
 
+    // show the widget again, recalculation of layout will matter again
+    // and so that the next step will scroll correctly to the first match
+    this._searchTarget.show();
+
     this._currentMatch = await this._stepNext(
       this._searchTarget.content.activeCell
     );
+    this._refreshCurrentCellEditor();
+
+    this._refreshCellsEditorsInBackground(this._cellsWithMatches);
 
     return allMatches;
   }
 
+  /**
+   * Gradually refresh cells in the background so that the user will not
+   * experience frozen interface, 5 cells at a time.
+   */
+  private _refreshCellsEditorsInBackground(cells: Cell[], n: number = 5) {
```

Comment:
In the ideal world, this would be implemented with a separate async worker, spawned at the beginning of the search and killed after the end of the search (i.e. closing the search widget AND finishing all the refresh tasks). This would allow modifying the list of the cells which need refreshing after each keystroke so that no cell is refreshed twice.

The set of cells to be refreshed could have an interface like {cell, priority} so that we could implement refreshing the cells closest to the user first (but I pass on this in this PR - I will have busy time now).